File: python-vj/vdj_monitor.py
#!/usr/bin/env python3
"""
VirtualDJ OSC Monitor - Minimal implementation

Subscribes to VDJ OSC for track info and is_audible state.
Implements get_playback() compatible with AppleScriptSpotifyMonitor.

Requirements: VirtualDJ PRO license, oscPort=9009, oscPortBack=9999
"""

import logging
import threading
import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

from osc import osc

logger = logging.getLogger(__name__)


@dataclass
class VDJDeckState:
    """Minimal state for a VirtualDJ deck."""
    deck: int = 1
    artist: str = ""
    title: str = ""
    album: str = ""
    key: str = ""
    duration_sec: float = 0.0
    is_playing: bool = False
    is_audible: bool = False  # VDJ: playing AND volume up
    position: float = 0.0  # 0.0-1.0
    bpm: float = 0.0
    beat_intensity: float = 0.0
    volume: float = 1.0
    loaded: bool = False
    _last_pos: float = 0.0  # For detecting position changes
    _last_pos_time: float = 0.0  # When position last changed


@dataclass
class VDJState:
    """Global VDJ state with two decks."""
    deck1: VDJDeckState = field(default_factory=lambda: VDJDeckState(deck=1))
    deck2: VDJDeckState = field(default_factory=lambda: VDJDeckState(deck=2))
    last_message_time: float = 0.0


class VDJMonitor:
    """Minimal VirtualDJ OSC Monitor."""
    
    monitor_key = "vdj_osc"
    monitor_label = "VirtualDJ (OSC)"
    
    # Subscribe to frequently-updating values to keep connection alive
    DECK_SUBS = [
        # Track info (changes on load)
        "get_title", "get_artist", "get_album", "get_key",
        "get_songlength", "get_bpm",
        # Playback state (changes frequently)
        "play", "is_audible", "song_pos", "volume", "loaded",
        "get_beat", "level",  # level changes constantly during playback
    ]
    
    # Global subscriptions (not per-deck)
    GLOBAL_SUBS = [
        "crossfader",  # Changes when mixing
        "masterlevel",  # Master output level
    ]
    
    TIMEOUT_SEC = 30.0  # VDJ only sends on change, so use longer timeout
    STARTUP_GRACE_SEC = 3.0  # Grace period after start before reporting unavailable
    QUERY_INTERVAL_SEC = 4.0  # Re-query VDJ to keep connection alive
    
    def __init__(self):
        self._state = VDJState()
        self._lock = threading.RLock()
        self._running = False
        self._start_time: float = 0.0
        self._last_query_time: float = 0.0
    
    def start(self) -> bool:
        """Start monitor, subscribe to VDJ OSC."""
        if self._running:
            return True
        try:
            if not osc.is_started:
                osc.start()
            osc.subscribe("/vdj/*", self._on_message)
            self._subscribe()
            self._running = True
            self._start_time = time.time()
            logger.info(f"[VDJMonitor] Started on port {osc.receive_port}")
            return True
        except Exception as e:
            logger.error(f"[VDJMonitor] Start failed: {e}")
            return False
    
    def stop(self):
        """Stop monitor."""
        if not self._running:
            return
        try:
            self._unsubscribe()
            osc.unsubscribe("/vdj/*", self._on_message)
        except Exception:
            pass
        self._running = False
    
    def get_playback(self) -> Optional[Dict[str, Any]]:
        """Get playback info for audible deck (TextlerEngine interface)."""
        # Auto-start on first call (lazy initialization)
        if not self._running:
            self.start()
        
        # Query VDJ for fresh data on each poll (VDJ doesn't push continuously)
        self._query_state()
        
        with self._lock:
            if not self._is_connected():
                return None
            deck = self._get_audible_deck()
            # Don't require 'loaded' flag - VDJ sends empty string for False
            # Instead, check if we have actual track info
            if not deck or (not deck.artist and not deck.title):
                return None
            return {
                'artist': deck.artist,
                'title': deck.title,
                'album': deck.album,
                'duration_ms': int(deck.duration_sec * 1000),
                'progress_ms': int(deck.position * deck.duration_sec * 1000),
                'is_playing': deck.is_playing,
                'source': 'vdj',
                'deck': deck.deck,
                'bpm': deck.bpm,
                'key': deck.key,
            }
    
    @property
    def status(self) -> Dict[str, Any]:
        """Monitor status for UI."""
        # Auto-start on first access
        if not self._running:
            self.start()
        with self._lock:
            connected = self._is_connected()
            deck = self._get_audible_deck()
            return {
                'available': connected,
                'status': 'connected' if connected else 'offline',
                'audible_deck': deck.deck if deck else None,
                'bpm': deck.bpm if deck else 0,
            }
    
    def _on_message(self, address: str, args: List[Any]):
        """Handle VDJ OSC message."""
        if not address.startswith("/vdj/"):
            return
        with self._lock:
            self._state.last_message_time = time.time()
        parts = address.split("/")
        if len(parts) >= 5 and parts[2] == "deck":
            try:
                deck_num = int(parts[3])
                verb = parts[4]
                self._handle_deck(deck_num, verb, args[0] if args else None)
            except (ValueError, IndexError):
                pass
    
    def _handle_deck(self, deck: int, verb: str, value: Any):
        """Update deck state from OSC."""
        if deck not in (1, 2):
            return
        with self._lock:
            d = self._state.deck1 if deck == 1 else self._state.deck2
            if verb == "get_title":
                d.title = str(value) if value else ""
            elif verb == "get_artist":
                d.artist = str(value) if value else ""
            elif verb == "get_album":
                d.album = str(value) if value else ""
            elif verb == "get_key":
                d.key = str(value) if value else ""
            elif verb == "get_songlength":
                d.duration_sec = float(value) if value else 0.0
            elif verb == "get_bpm":
                d.bpm = float(value) if value else 0.0
            elif verb == "get_beat":
                d.beat_intensity = float(value) if value else 0.0
            elif verb == "play":
                # VDJ sends "1" for playing, "" for stopped
                d.is_playing = value not in (None, "", 0, "0", False)
            elif verb == "is_audible":
                d.is_audible = value not in (None, "", 0, "0", False)
            elif verb == "song_pos":
                new_pos = float(value) if value else 0.0
                now = time.time()
                # Infer playing from position changing (VDJ doesn't always send play state)
                if abs(new_pos - d._last_pos) > 0.001:
                    d._last_pos = new_pos
                    d._last_pos_time = now
                    d.is_playing = True  # Position changed = playing
                    d.is_audible = d.volume > 0.1  # Infer audible from volume
                d.position = new_pos
            elif verb == "volume":
                d.volume = float(value) if value else 1.0
            elif verb == "loaded":
                d.loaded = value not in (None, "", 0, "0", False)
    
    def _query_state(self):
        """Query VDJ for current state (VDJ only responds to queries, not push)."""
        # Query essential values for both decks
        for deck in (1, 2):
            osc.vdj.send(f"/vdj/query/deck/{deck}/song_pos")
            osc.vdj.send(f"/vdj/query/deck/{deck}/get_title")
            osc.vdj.send(f"/vdj/query/deck/{deck}/get_artist")
            osc.vdj.send(f"/vdj/query/deck/{deck}/volume")
    
    def _subscribe(self):
        """Subscribe to VDJ OSC updates (for change notifications)."""
        # Per-deck subscriptions
        for deck in (1, 2):
            for verb in self.DECK_SUBS:
                osc.vdj.send(f"/vdj/subscribe/deck/{deck}/{verb}")
        # Global subscriptions (update frequently = keep-alive)
        for verb in self.GLOBAL_SUBS:
            osc.vdj.send(f"/vdj/subscribe/{verb}")
        # Query initial state
        for deck in (1, 2):
            for verb in self.DECK_SUBS:
                osc.vdj.send(f"/vdj/query/deck/{deck}/{verb}")
    
    def _unsubscribe(self):
        """Unsubscribe from VDJ OSC."""
        for deck in (1, 2):
            for verb in self.DECK_SUBS:
                osc.vdj.send(f"/vdj/unsubscribe/deck/{deck}/{verb}")
        for verb in self.GLOBAL_SUBS:
            osc.vdj.send(f"/vdj/unsubscribe/{verb}")
    
    def _get_audible_deck(self) -> Optional[VDJDeckState]:
        """Return audible deck (VDJ is_audible = playing + volume up)."""
        d1, d2 = self._state.deck1, self._state.deck2
        # Check is_audible first (VDJ: playing AND volume up)
        if d1.is_audible and not d2.is_audible:
            return d1
        if d2.is_audible and not d1.is_audible:
            return d2
        if d1.is_audible and d2.is_audible:
            return d1  # Both audible, pick deck 1
        # Fallback: any deck with track info (artist or title present)
        d1_has_track = bool(d1.artist or d1.title)
        d2_has_track = bool(d2.artist or d2.title)
        if d1_has_track and not d2_has_track:
            return d1
        if d2_has_track and not d1_has_track:
            return d2
        # Both have tracks or neither - prefer deck 1
        return d1 if d1_has_track else None
    
    def _is_connected(self) -> bool:
        """Check if VDJ responded recently, with startup grace period."""
        # During startup grace period, assume connected (waiting for first message)
        if self._start_time > 0:
            elapsed = time.time() - self._start_time
            if elapsed < self.STARTUP_GRACE_SEC:
                return True  # Still in grace period
        t = self._state.last_message_time
        return t > 0 and (time.time() - t) < self.TIMEOUT_SEC

File: swift-vj/Sources/SwiftVJCore/Adapters/VDJMonitor.swift
// VDJMonitor - Monitor VirtualDJ playback via OSC
// Following Grokking Simplicity: this is an action (side effects)

import Foundation
import OSCKit

/// Deck information from VirtualDJ
public struct VDJDeck: Sendable, Equatable {
    public let deckNumber: Int
    public let artist: String
    public let title: String
    public let album: String
    public let bpm: Double
    public let position: Double      // seconds
    public let duration: Double      // seconds
    public let isPlaying: Bool       // VDJ "play" state
    public let isAudible: Bool       // VDJ "is_audible" = playing AND volume up (KEY for deck selection!)
    public let isMaster: Bool
    public let volume: Double        // 0.0-1.0
    
    public init(
        deckNumber: Int,
        artist: String = "",
        title: String = "",
        album: String = "",
        bpm: Double = 0,
        position: Double = 0,
        duration: Double = 0,
        isPlaying: Bool = false,
        isAudible: Bool = false,
        isMaster: Bool = false,
        volume: Double = 1.0
    ) {
        self.deckNumber = deckNumber
        self.artist = artist
        self.title = title
        self.album = album
        self.bpm = bpm
        self.position = position
        self.duration = duration
        self.isPlaying = isPlaying
        self.isAudible = isAudible
        self.isMaster = isMaster
        self.volume = volume
    }
    
    /// Track key for change detection
    public var trackKey: String {
        "\(artist.lowercased())|\(title.lowercased())"
    }
    
    /// Check if deck has a track loaded (artist OR title - matching Python)
    public var hasTrack: Bool {
        !artist.isEmpty || !title.isEmpty
    }
}

/// VirtualDJ playback state (both decks)
public struct VDJPlayback: Sendable, Equatable {
    public let deck1: VDJDeck
    public let deck2: VDJDeck
    public let masterBPM: Double
    public let crossfader: Double  // -1.0 (left) to 1.0 (right)
    
    public init(
        deck1: VDJDeck = VDJDeck(deckNumber: 1),
        deck2: VDJDeck = VDJDeck(deckNumber: 2),
        masterBPM: Double = 0,
        crossfader: Double = 0
    ) {
        self.deck1 = deck1
        self.deck2 = deck2
        self.masterBPM = masterBPM
        self.crossfader = crossfader
    }
    
    /// Get the currently audible deck - the LOUDER deck with a track
    /// Priority: isAudible flag → volume comparison → hasTrack fallback
    public var audibleDeck: VDJDeck? {
        // 1. Check is_audible first (VDJ: playing AND volume up) - if VDJ sends it
        if deck1.isAudible && !deck2.isAudible { return deck1 }
        if deck2.isAudible && !deck1.isAudible { return deck2 }
        
        // 2. Both have tracks? Pick the LOUDER one by volume
        //    This handles crossfade: louder deck = the one coming in
        if deck1.hasTrack && deck2.hasTrack {
            // Significant volume difference (>10%) → pick louder
            if deck2.volume > deck1.volume + 0.1 { return deck2 }
            if deck1.volume > deck2.volume + 0.1 { return deck1 }
            // Similar volume → prefer deck 1 (like Python)
            return deck1
        }
        
        // 3. Only one has a track
        if deck1.hasTrack && !deck2.hasTrack { return deck1 }
        if deck2.hasTrack && !deck1.hasTrack { return deck2 }
        
        // 3. Both have tracks or neither - prefer deck 1 (like Python)
        return deck1.hasTrack ? deck1 : nil
    }
}

/// Error types for VDJ monitoring
public enum VDJMonitorError: Error, Equatable {
    case vdjNotRunning
    case oscError(String)
    case timeout
}

/// Handler for VDJ track changes
public typealias VDJTrackHandler = @Sendable (VDJDeck) -> Void

/// Handler for VDJ position updates
public typealias VDJPositionHandler = @Sendable (Int, Double) -> Void  // deck, position

/// Monitor VirtualDJ playback via OSC subscriptions
/// 
/// VDJ sends OSC messages on these addresses:
/// - /deck/1/artist, /deck/1/title, /deck/1/album
/// - /deck/1/get_bpm, /deck/1/get_position, /deck/1/get_duration
/// - /deck/1/play, /deck/1/masterdeck, /deck/1/volume
/// - /crossfader, /masterdeck_bpm
public actor VDJMonitor {
    
    // MARK: - Configuration
    
    public static let vdjOSCPort: UInt16 = 9009
    
    // MARK: - State
    
    private var deck1: VDJDeck = VDJDeck(deckNumber: 1)
    private var deck2: VDJDeck = VDJDeck(deckNumber: 2)
    private var masterBPM: Double = 0
    private var crossfader: Double = 0
    
    // Position tracking for inferring audibility (Python: _last_pos, _last_pos_time)
    private var deck1LastPos: Double = 0
    private var deck2LastPos: Double = 0
    
    private let health: ServiceHealth
    private var lastUpdate: Date = .distantPast
    private var isSubscribed = false
    
    // Callbacks
    private var trackHandlers: [VDJTrackHandler] = []
    private var positionHandlers: [VDJPositionHandler] = []
    
    public init() {
        self.health = ServiceHealth(name: "VirtualDJ")
    }
    
    // MARK: - VDJ OSC Subscription Commands
    
    /// Deck subscriptions (VDJ pushes on change)
    private static let deckSubscriptions = [
        "get_title", "get_artist", "get_album", "get_key",
        "get_songlength", "get_bpm", "loaded"
    ]
    
    /// Global subscriptions
    private static let globalSubscriptions = [
        "crossfader"
    ]
    
    /// Deck queries (poll for current values)
    /// Include metadata fields since VDJ may not push on initial subscribe
    private static let deckQueries = [
        "get_title", "get_artist", "get_album", "get_bpm", "get_songlength",
        "song_pos", "volume", "is_audible", "play"
    ]
    
    /// Subscribe to VDJ OSC events via OSCHub
    /// Call this after wiring handleOSC to the hub
    public func subscribe(using hub: OSCHub) throws {
        guard !isSubscribed else { return }
        
        var sentCount = 0
        // Subscribe to discrete events (VDJ pushes these)
        for deck in [1, 2] {
            for verb in Self.deckSubscriptions {
                let addr = "/vdj/subscribe/deck/\(deck)/\(verb)"
                try hub.sendToVDJ(addr)
                sentCount += 1
            }
        }
        
        // Global subscriptions
        for verb in Self.globalSubscriptions {
            try hub.sendToVDJ("/vdj/subscribe/\(verb)")
            sentCount += 1
        }
        
        print("📤 VDJMonitor: sent \(sentCount) subscribe commands to VDJ port 9009")
        isSubscribed = true
    }
    
    /// Query VDJ for current state (poll)
    public func query(using hub: OSCHub) throws {
        for deck in [1, 2] {
            for verb in Self.deckQueries {
                try hub.sendToVDJ("/vdj/query/deck/\(deck)/\(verb)")
            }
        }
    }
    
    // MARK: - Public API
    
    /// Get current playback state
    public func getPlayback() -> VDJPlayback {
        VDJPlayback(
            deck1: deck1,
            deck2: deck2,
            masterBPM: masterBPM,
            crossfader: crossfader
        )
    }
    
    /// Get the currently audible track
    public func getAudibleTrack() -> VDJDeck? {
        getPlayback().audibleDeck
    }
    
    /// Subscribe to track changes
    public func onTrackChange(_ handler: @escaping VDJTrackHandler) {
        trackHandlers.append(handler)
    }
    
    /// Subscribe to position updates
    public func onPositionUpdate(_ handler: @escaping VDJPositionHandler) {
        positionHandlers.append(handler)
    }
    
    /// Process incoming OSC message from VDJ
    public func handleOSC(address: String, values: [any OSCValue]) {
        // Parse deck number from address
        guard let deckNum = parseDeckNumber(from: address) else {
            handleGlobalOSC(address: address, values: values)
            return
        }
        
        // Get first value - pass raw for proper boolean parsing
        let firstValue = values.first
        let stringValue = (firstValue as? String) ?? ""
        let floatValue = firstValue.flatMap { extractFloat(from: $0) } ?? 0
        let boolValue = extractBool(from: firstValue)
        
        // Update deck state based on address suffix
        let suffix = address.components(separatedBy: "/").last ?? ""
        updateDeckValue(deckNum: deckNum, suffix: suffix, stringValue: stringValue, floatValue: floatValue, boolValue: boolValue)
        
        lastUpdate = Date()
        Task { await health.markAvailable(message: "VDJ connected") }
    }
    
    /// Check if monitor is receiving data
    public func isReceiving() -> Bool {
        Date().timeIntervalSince(lastUpdate) < 5.0
    }
    
    /// Get service health status
    public func status() async -> [String: Any] {
        await health.status()
    }
    
    // MARK: - Private
    
    private func parseDeckNumber(from address: String) -> Int? {
        // Python: parts = address.split("/"), if parts[2] == "deck": deck_num = int(parts[3])
        // For /vdj/deck/2/is_audible: parts = ["", "vdj", "deck", "2", "is_audible"]
        // For /deck/2/is_audible: parts = ["", "deck", "2", "is_audible"]
        let parts = address.split(separator: "/").map { String($0) }
        
        // Find "deck" in parts and get the next element as deck number
        for (index, part) in parts.enumerated() {
            if part == "deck", index + 1 < parts.count {
                if let deckNum = Int(parts[index + 1]) {
                    return deckNum
                }
            }
        }
        return nil
    }
    
    private func handleGlobalOSC(address: String, values: [any OSCValue]) {
        let floatValue = values.first.flatMap { extractFloat(from: $0) } ?? 0
        
        switch address {
        case "/crossfader":
            crossfader = floatValue * 2 - 1  // Convert 0-1 to -1..1
        case "/masterdeck_bpm", "/master_bpm":
            masterBPM = floatValue
        default:
            break
        }
    }
    
    private func updateDeckValue(deckNum: Int, suffix: String, stringValue: String, floatValue: Double, boolValue: Bool) {
        let oldDeck = deckNum == 1 ? deck1 : deck2
        let oldTrackKey = oldDeck.trackKey
        let lastPos = deckNum == 1 ? deck1LastPos : deck2LastPos
        
        var deck = oldDeck
        
        switch suffix {
        case "artist", "get_artist":
            deck = VDJDeck(
                deckNumber: deckNum, artist: stringValue, title: deck.title,
                album: deck.album, bpm: deck.bpm, position: deck.position,
                duration: deck.duration, isPlaying: deck.isPlaying,
                isAudible: deck.isAudible, isMaster: deck.isMaster, volume: deck.volume
            )
        case "title", "get_title":
            deck = VDJDeck(
                deckNumber: deckNum, artist: deck.artist, title: stringValue,
                album: deck.album, bpm: deck.bpm, position: deck.position,
                duration: deck.duration, isPlaying: deck.isPlaying,
                isAudible: deck.isAudible, isMaster: deck.isMaster, volume: deck.volume
            )
        case "album", "get_album":
            deck = VDJDeck(
                deckNumber: deckNum, artist: deck.artist, title: deck.title,
                album: stringValue, bpm: deck.bpm, position: deck.position,
                duration: deck.duration, isPlaying: deck.isPlaying,
                isAudible: deck.isAudible, isMaster: deck.isMaster, volume: deck.volume
            )
        case "get_bpm", "bpm":
            deck = VDJDeck(
                deckNumber: deckNum, artist: deck.artist, title: deck.title,
                album: deck.album, bpm: floatValue, position: deck.position,
                duration: deck.duration, isPlaying: deck.isPlaying,
                isAudible: deck.isAudible, isMaster: deck.isMaster, volume: deck.volume
            )
        case "get_position", "position", "song_pos":
            // song_pos comes as 0.0-1.0 ratio, convert to seconds using duration
            let positionInSeconds = floatValue <= 1.0 ? floatValue * deck.duration : floatValue
            
            // Python: Infer playing from position changing (VDJ doesn't always send play state)
            let positionChanged = abs(floatValue - lastPos) > 0.001
            var inferredPlaying = deck.isPlaying
            var inferredAudible = deck.isAudible
            
            if positionChanged {
                inferredPlaying = true  // Position changed = playing
                inferredAudible = deck.volume > 0.1  // Infer audible from volume
                // Update last pos
                if deckNum == 1 { deck1LastPos = floatValue } else { deck2LastPos = floatValue }
            }
            
            deck = VDJDeck(
                deckNumber: deckNum, artist: deck.artist, title: deck.title,
                album: deck.album, bpm: deck.bpm, position: positionInSeconds,
                duration: deck.duration, isPlaying: inferredPlaying,
                isAudible: inferredAudible, isMaster: deck.isMaster, volume: deck.volume
            )
            // Notify position handlers
            for handler in positionHandlers {
                handler(deckNum, positionInSeconds)
            }
        case "get_duration", "duration", "get_songlength", "songlength":
            deck = VDJDeck(
                deckNumber: deckNum, artist: deck.artist, title: deck.title,
                album: deck.album, bpm: deck.bpm, position: deck.position,
                duration: floatValue, isPlaying: deck.isPlaying,
                isAudible: deck.isAudible, isMaster: deck.isMaster, volume: deck.volume
            )
        case "play":
            // VDJ sends "1" for playing, "" for stopped - use boolValue (Python-style parsing)
            deck = VDJDeck(
                deckNumber: deckNum, artist: deck.artist, title: deck.title,
                album: deck.album, bpm: deck.bpm, position: deck.position,
                duration: deck.duration, isPlaying: boolValue,
                isAudible: deck.isAudible, isMaster: deck.isMaster, volume: deck.volume
            )
        case "is_audible":
            // VDJ "is_audible" = playing AND volume up - use boolValue (Python-style parsing)
            deck = VDJDeck(
                deckNumber: deckNum, artist: deck.artist, title: deck.title,
                album: deck.album, bpm: deck.bpm, position: deck.position,
                duration: deck.duration, isPlaying: deck.isPlaying,
                isAudible: boolValue, isMaster: deck.isMaster, volume: deck.volume
            )
        case "masterdeck":
            deck = VDJDeck(
                deckNumber: deckNum, artist: deck.artist, title: deck.title,
                album: deck.album, bpm: deck.bpm, position: deck.position,
                duration: deck.duration, isPlaying: deck.isPlaying,
                isAudible: deck.isAudible, isMaster: boolValue, volume: deck.volume
            )
        case "volume":
            deck = VDJDeck(
                deckNumber: deckNum, artist: deck.artist, title: deck.title,
                album: deck.album, bpm: deck.bpm, position: deck.position,
                duration: deck.duration, isPlaying: deck.isPlaying,
                isAudible: deck.isAudible, isMaster: deck.isMaster, volume: floatValue
            )
        case "loaded":
            // Track loaded state using boolValue (Python-style parsing)
            return  // loaded doesn't update deck state directly
        default:
            return
        }
        
        // Store updated deck
        if deckNum == 1 {
            deck1 = deck
        } else {
            deck2 = deck
        }
        
        // Check for track change
        if deck.trackKey != oldTrackKey && deck.hasTrack {
            for handler in trackHandlers {
                handler(deck)
            }
        }
    }
    
    private func extractFloat(from value: any OSCValue) -> Double? {
        if let f = value as? Float32 { return Double(f) }
        if let f = value as? Float64 { return f }
        if let i = value as? Int32 { return Double(i) }
        if let s = value as? String { return Double(s) }
        return nil
    }
    
    /// Parse boolean the way Python does: value not in (None, "", 0, "0", False)
    /// VDJ sends "1" for true, "" for false - this handles all cases
    private func extractBool(from value: (any OSCValue)?) -> Bool {
        guard let value = value else { return false }
        
        // Check string values first (VDJ often sends "1" or "")
        if let s = value as? String {
            return !s.isEmpty && s != "0" && s.lowercased() != "false"
        }
        // Check numeric values
        if let f = value as? Float32 { return f != 0 }
        if let f = value as? Float64 { return f != 0 }
        if let i = value as? Int32 { return i != 0 }
        if let b = value as? Bool { return b }
        
        return false
    }
}
